data_diff/databases/clickhouse.py

```diff
@@ -0,0 +1,153 @@
+
+        self._args = kw
+        # In Clickhouse database and schema are the same
+        self.default_schema = kw['database']
+
+    def create_connection(self):
+        clickhouse = import_clickhouse()
+        try:
+            connection = SingleConnection(**self._args)
+            return connection
+        except clickhouse.OperationError as e:
+            raise ConnectError(*e.args) from e
+
+    def _parse_type_repr(self, type_repr: str) -> Optional[Type[ColType]]:
+        nullable_prefix = 'Nullable'
+        if type_repr.lower().startswith(nullable_prefix.lower()):
+            type_repr = type_repr[len(nullable_prefix):].lstrip('(').rstrip(')')
+
+        if type_repr.startswith('Decimal'):
+            type_repr = 'Decimal'
+        elif type_repr.startswith('FixedString'):
+            type_repr = 'FixedString'
+        elif type_repr.startswith('DateTime64'):
+            type_repr = 'DateTime64'
+
+        return self.TYPE_CLASSES.get(type_repr)
+
+    def quote(self, s: str) -> str:
+        return f'"{s}"'
+
+    def md5_to_int(self, s: str) -> str:
+        substr_idx = 1 + MD5_HEXDIGITS - CHECKSUM_HEXDIGITS
+        return f'reinterpretAsUInt128(reverse(unhex(lowerUTF8(substr(hex(MD5({s})), {substr_idx})))))'
+
+    def to_string(self, s: str) -> str:
+        return f"toString({s})"
+
+    def normalize_timestamp(self, value: str, coltype: TemporalType) -> str:
+        if coltype.rounds:
+            prec = coltype.precision
+            timestamp = f'toDateTime64(round(toUnixTimestamp64Micro(toDateTime64({value}, 6)) / 1000000, {prec}), 6)'
+            return self.to_string(timestamp)
+        else:
+            fractional = f'toUnixTimestamp64Micro(toDateTime64({value}, 6)) % 1000000'
+            return f"formatDateTime({value}, '%Y-%m-%d %H:%M:%S') || '.' || {self.to_string(fractional)}"
+
+    def _convert_db_precision_to_digits(self, p: int) -> int:
+        # Done the same as for PostgreSQL but need to rewrite in another way
+        # because it does not help for float with a big integer part.
+        return super()._convert_db_precision_to_digits(p) - 2
+
+    def normalize_number(self, value: str, coltype: FractionalType) -> str:
+        # If a decimal value has trailing zeros in a fractional part, when casting to string they are dropped.
+        # For example:
+        #   select toString(toDecimal128(1.10, 2));  -- the result is 1.1
+        #   select toString(toDecimal128(1.00, 2)); -- the result is 1
+        # So, we should use some custom approach to save these trailing zeros.
+        # To avoid it, we can add a small value like 0.000001 to prevent dropping of zeros from the end when casting.
+        # For examples above it looks like:
+        #   select toString(toDecimal128(1.10, 2 + 1) + toDecimal128(0.001, 3)); -- the result is 1.101
+        # After that, cut an extra symbol from the string, i.e. 1.101 -> 1.10
+        # So, the algorithm is:
```

Comment:
Why is this better than calling RPAD, once with "." and once with "0" ?

It sounds like your method might lead to all kinds of strange precision errors.

Reply:
> Why is this better than calling RPAD, once with "." and once with "0" ?

It's a problem to obtain a number of digits (it may vary in contrast to timestamp format) to apply `rpad` here.

> It sounds like your method might lead to all kinds of strange precision errors.
Does it? I have just added a small number with a higher precision here, and do not use it to round something.

Reply:
I can't say for sure. It just feels unstable, but maybe I'm worried for nothing.

I'll give it a deeper look later.

Reply:
To the current moment I see only one corner case, namely, if a customer's number type has precision higher than 38 digits (for example, they can use `Decimal256`, or `Decimal(100, 40)` etc). Casting to `Decimal128(39)`is not possible. However, I'm note sure that people has such high precision.

Also I tried to rewrite it using RPAD but found with next problem. When we do not add a little value, casting of float to decimal has an unpleasant side effect. For example, let's we want to cast to precision equaled 5:
```
with tmp as (
    select toFloat64(0.001201923076923077) as col
)
select
    col,
    round(col, 5),
    toDecimal128(round(col, 5), 5) as precision_5,
    toDecimal128(round(col, 5), 6) as precision_6
from tmp
order by col;
```
Result:
```
0.001201923076923077, 0.0012, 0.00119, 0.001200
```
As you can see, the number was corrupted for precision for 5.